// @oasislabs/parcel-sdk/lib/src/identity/keypair.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const nacl = __importStar(require("tweetnacl"));
const oasis_std_1 = require("oasis-std");
/**
 * An ed25519 key pair.
 */
class KeyPair {
    constructor(publicKey, secretKey) {
        this.publicKey = publicKey;
        this.secretKey = secretKey;
    }
    static generate() {
        const { publicKey, secretKey } = nacl.sign.keyPair();
        return new KeyPair(publicKey, secretKey);
    }
    /**
     * @returns the (ed25519) signature of this key over `message`.
     */
    sign(message) {
        return nacl.sign.detached(message, this.secretKey);
    }
    toJSON() {
        return {
            publicKey: oasis_std_1.encodeHex(this.publicKey),
            secretKey: oasis_std_1.encodeHex(this.secretKey),
        };
    }
    static fromJSON(key, value) {
        if (key === 'publicKey' || key === 'secretKey') {
            return oasis_std_1.decodeHex(value);
        }
        else {
            return value;
        }
    }
}
exports.KeyPair = KeyPair;
KeyPair.PUBLIC_KEY_LEN = nacl.sign.publicKeyLength;
KeyPair.SECRET_KEY_LEN = nacl.sign.secretKeyLength;
